### PR TITLE
Assignment 3: @Clockknight

### PR DESCRIPTION
**Pull request by Malcolm-Shepherd**

### CODE REVIEW

**Review by BenFristadSchool — COMMENTED** [review]
I will review this

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            JokeIn = jokeIn;
+            JokeOut = jokeOut;
+        }
+        public void TellJoke()
+        {
+            string joke;
+            do { joke = this.JokeIn.GetJoke(); }
+            while (joke.Contains("Chuck Norris"));
+
+            this.JokeOut.JokeToConsole(joke);
```
Comment (BenFristadSchool) [inline CanHazFunny/CanHazFunny/Jester.cs:47]: writes joke to console

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 2]
```diff
@@ -1,10 +1,62 @@
...
+using System;
+using System.IO;
+using System.Linq;
 
 namespace CanHazFunny.Tests
 {
     [TestClass]
     public class JesterTests
     {
-
```
Comment (BenFristadSchool) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:8]: code coverage is above 90%

File: CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj [hunk 3]
```diff
@@ -1,8 +1,12 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>net6.0</TargetFramework>
-
+	  <TargetFramework>net7.0</TargetFramework>
+	  <LangVersion>11.0</LangVersion>
+	  <Nullable>enable</Nullable>
```
Comment (BenFristadSchool) [inline CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj:6]: nullability is enabled

File: CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj [hunk 4]
```diff
@@ -1,8 +1,12 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>net6.0</TargetFramework>
-
+	  <TargetFramework>net7.0</TargetFramework>
+	  <LangVersion>11.0</LangVersion>
```
Comment (BenFristadSchool) [inline CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj:5]: lang is set to current version

File: CanHazFunny/CanHazFunny/CanHazFunny.csproj [hunk 5]
```diff
@@ -2,7 +2,12 @@
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>net6.0</TargetFramework>
+    <TargetFramework>net7.0</TargetFramework>
```
Comment (BenFristadSchool) [inline CanHazFunny/CanHazFunny/CanHazFunny.csproj:5]: target framework is set to current version

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 6]
```diff
@@ -1,10 +1,62 @@
...
+        //does jester implement jokeservice
+        //does jester implement jokewriter
+        [TestMethod]
+        public void Constructor_Interfaces_as_Dependencies()
+        {
+            JokeService testService = new JokeService();
+            JokeWriter testWriter = new JokeWriter();
+            Jester tester = new Jester(testService, testWriter);
+
+            Assert.AreEqual<JokeService>(testService, tester.JokeIn);
```
Comment (BenFristadSchool) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:24]: generics are used for comparisons

**Review by BenFristadSchool — COMMENTED** [review]
I could not find any flaws in the assignment. I just verified everything was done correctly. Great job! The only thing I would check is that the xml is consistent between CanHazFunny.csproj and CanHazFunny.Tests.csproj

**Comment by mhtran23** [comment]
For this assignment do the following:

Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
Fundamentals
Be sure you enable:
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Everything works and looks great!

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs (hunk 6 above)
Comment (Clockknight) [reply CanHazFunny/CanHazFunny.Tests/JesterTests.cs:24]: Thanks so much for the overview! Will implement these today.

**Comment by Clockknight** [comment]
Appreciate the feedback, @mhtran23 :)

**Comment by Clockknight** [comment]
@BenFristadSchool Thanks so much for the overview! Will implement these today.

**Comment by Clockknight** [comment]
@BenFristadSchool Thanks so much for the overview!

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            {
+                return _JokeOut!;
+            }
+            set
+            {
+                _JokeOut = value ?? throw new ArgumentNullException(nameof(value));
+            }
+        }
+
+        public Jester(JokeService jokeIn, JokeWriter jokeOut)
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:36]:
```suggestion
        public Jester(IJokeService jokeService, IJokeWriter jokeWriter) 
```
These should be your interfaces. Also, keep the naming the same, just vary the casing throughout your code.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+        public JokeService JokeIn { 
+            get {
+                return _JokeIn!;
+            }
+            set {
+                _JokeIn = value ?? throw new ArgumentNullException(nameof(value));
+            } 
+        }
+        public JokeWriter JokeOut
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:24]:
```suggestion
        public IJokeWriter JokeWriter
```

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+
+namespace CanHazFunny
+{
+    public class Jester
+    {
+        private JokeService? _JokeIn; 
+        private JokeWriter? _JokeOut;
+
+        public JokeService JokeIn {
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:16]:
```suggestion
        public IJokeService JokeService { 
```

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+        public Jester(JokeService jokeIn, JokeWriter jokeOut) 
+        {
+            JokeIn = jokeIn;
+            JokeOut = jokeOut;
+        }
+        public void TellJoke()
+        {
+            string joke;
+            do { joke = this.JokeIn.GetJoke(); }
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:44]:
```suggestion
            do 
            { 
                joke = this.JokeIn.GetJoke(); 
             }
```
this should be multiple lines, just out of convention

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            while (joke.Contains("Chuck Norris"));
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:45]: Consider checking lower case here as well!

**Pull request closed without merging**